### PR TITLE
Construct addresses using MultiAddr, defaulting to IPv6

### PR DESCRIPTION
**Pull request by DougAnderson444**
### Rust-Peer (server) improvements

I was working in my own repo, but I figure I would start using this repo instead of building in a parallel silo. Here are some changes I would propose for this repo.

I know this Rust server is meant for the cloud, but seeing as I am a big fan of "running a peer at home" this PR started out as changing the default IP to v6 so home servers are easier to connect, and then I added a few more cleanup items along the way. EC2 should run on IPv6 addresses just fine too, even though it has a public IP.

### Summary of changes

- use IPv6 by default, as they offer globally available IP addresses (IPv4 is so 1998)
- ~~note that `ipv6.is_global()` is [only available](https://github.com/rust-lang/rust/issues/27709) on [nightly](https://doc.rust-lang.org/std/net/enum.IpAddr.html#method.is_global), I hope that's not a problem? I added a toolchain config for that. If not we can code around it with logic I suppose...~~ **use `!ip.is_looback()` to only add external addresses**
- use address variables (`address_webrtc` & `address_quic`)
- only add globally available external addresses to the `swarm`
- construct address by building with `MultiAddr.with(..)`
- move constants to beginning of `main.rs`
- match on `Opt` `listen_address` (String) to ensure it's a valid IP address (`IpAddr`)
- cleanup/remove unused deps, & ping

### Remaining work

- I left the dependencies as-is for now, whenever I moved off the branches things broke, and I wanted to get these changes in incrementally. It's late at night now, so perhaps we can migrate/patch the deps in another PR?

### CODE REVIEW

**Review by p-shahi — APPROVED** [review]

**Comment by p-shahi** [comment]
Thank you @DougAnderson444 🙌 

I'm guessing you tested locally?

> (IPv4 is so 1998)

@marten-seemann would agree 😂  , he pointed out the same in the go-peer pull request. Yeah let's enable v6!
I'll get the v6 elastic IPs set up.

> I left the dependencies as-is for now, whenever I moved off the branches things broke, and I wanted to get these changes in incrementally. It's late at night now, so perhaps we can migrate/patch the deps in another PR?

https://github.com/libp2p/rust-libp2p/pull/3625 was merged into `master` so we're super close to moving to not relying on my forked repo. However, I still need to apply https://github.com/p-shahi/rust-libp2p/commit/201c7a88a82983bb35b409f55f85f9bf3c1947b0 onto the master branch to get it working. I'll take care of this in a separate PR

**Comment by DougAnderson444** [comment]
> 
> I'm guessing you tested locally?
> 

Yes I tested locally, works like a charm. I didn't test on EC2 of course, I'd have to leave that to you.
 
> [libp2p/rust-libp2p#3625](https://github.com/libp2p/rust-libp2p/pull/3625) was merged into `master` so we're super close to moving to not relying on my forked repo. However, I still need to apply [p-shahi/rust-libp2p@201c7a8](https://github.com/p-shahi/rust-libp2p/commit/201c7a88a82983bb35b409f55f85f9bf3c1947b0) onto the master branch to get it working. I'll take care of this in a separate PR

Sounds great! Yes I've been watching that PR as well, glad to see mergify finally did it's thing.

**Comment by DougAnderson444** [comment]
> However, I still need to apply [p-shahi/rust-libp2p@201c7a8](https://github.com/p-shahi/rust-libp2p/commit/201c7a88a82983bb35b409f55f85f9bf3c1947b0) onto the master branch to get it working.

Ah, I see what you mean, `WebRTCDirect` changes were merged here https://github.com/multiformats/rust-multiaddr/pull/78 by @mxinden but `v0.18.0` hasn't yet been released, nor used in libp2p [core](https://github.com/libp2p/rust-libp2p/blob/4b41f5a9942add3f7a815bd2b79cb82eaed1ee05/core/Cargo.toml#L24)

**Comment by p-shahi** [comment]
Let's hold off on merging this for bit

**Comment by p-shahi** [comment]
@DougAnderson444 can you fix the merge conflicts? and then I'll merge

**Comment by DougAnderson444** [comment]
Yes merging now. 

Is there any particular reason why `webrtc` (webrtc-rs) is a dependency in Cargo.toml? Libp2p-webrtc should provide everything rust needs for the transport

**Comment by p-shahi** [comment]
> Is there any particular reason why webrtc (webrtc-rs) is a dependency in Cargo.toml? Libp2p-webrtc should provide everything rust needs for the transport

I think you can remove that and multiaddr

File: rust-peer/rust-toolchain.toml
Comment (p-shahi) [inline rust-peer/rust-toolchain.toml:1]: maybe add a comment similar to what you mentioned here:

> note that ipv6.is_global() is https://github.com/rust-lang/rust/issues/27709 on [nightly](https://doc.rust-lang.org/std/net/enum.IpAddr.html#method.is_global), I hope that's not a problem? I added a toolchain config for that. If not we can code around it with logic I suppose...

Comment (DougAnderson444) [reply rust-peer/rust-toolchain.toml:1]: Added.

Comment (DougAnderson444) [reply rust-peer/rust-toolchain.toml:1]: Actually, I think we'll just use `!ip.is_loopback()` so we don't have to set the toolchain at all. All we really want to do is make sure we're not adding a local IP anyway, so it works

**Comment by DougAnderson444** [comment]
I have removed the dead dependencies, and removed the dep on the rust toolchain channel (no more need for nightly to use `!ip.is_looback()`

**Comment by maschad** [comment]
Thanks for this @DougAnderson444 🚀  

It seems there is still a conflict in the [Cargo.toml](https://github.com/libp2p/universal-connectivity/pull/46/files#diff-3e777d5c562f37faa9d203ae6f7a539d093957c723720c7cb8d90864fd091675) , once that's resolved we should be good to land this.

**Comment by DougAnderson444** [comment]
@maschad all deconflicted over here.

**Review by maschad — APPROVED** [review]
LGTM

**Pull request merged**
